src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/ControllerProvider.cs

```diff
@@ -0,0 +1,83 @@
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp
+{
+    [Export]
+    [Shared]
+    internal class ControllerProvider : ForegroundThreadAffinitizedObject
+    {
+        private static readonly object s_controllerPropertyKey = new();
+
+        private readonly IIntelliSensePresenter<ISignatureHelpPresenterSession, ISignatureHelpSession> _signatureHelpPresenter;
+        private readonly IAsynchronousOperationListener _listener;
+        private readonly IAsyncCompletionBroker _completionBroker;
+        private readonly IList<Lazy<ISignatureHelpProvider, OrderableLanguageMetadata>> _signatureHelpProviders;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public ControllerProvider(
+            IThreadingContext threadingContext,
+            [ImportMany] IEnumerable<Lazy<ISignatureHelpProvider, OrderableLanguageMetadata>> signatureHelpProviders,
+            [ImportMany] IEnumerable<Lazy<IIntelliSensePresenter<ISignatureHelpPresenterSession, ISignatureHelpSession>, OrderableMetadata>> signatureHelpPresenters,
+            IAsyncCompletionBroker completionBroker,
+            IAsynchronousOperationListenerProvider listenerProvider)
+            : base(threadingContext)
+        {
+            _signatureHelpPresenter = ExtensionOrderer.Order(signatureHelpPresenters).Select(lazy => lazy.Value).FirstOrDefault();
+            _listener = listenerProvider.GetListener(FeatureAttribute.SignatureHelp);
+            _completionBroker = completionBroker;
+            _signatureHelpProviders = ExtensionOrderer.Order(signatureHelpProviders);
+        }
+
+        public Controller? GetController(ITextView textView, ITextBuffer subjectBuffer)
+        {
+            AssertIsForeground();
+
+            if (textView.TryGetPerSubjectBufferProperty<Controller, ITextView>(subjectBuffer, s_controllerPropertyKey, out var controller))
+                return controller;
+
+            // If we don't have a presenter, then there's no point in us even being involved.
+            if (_signatureHelpPresenter == null)
+            {
+                return null;
+            }
+
+            return GetControllerSlow(textView, subjectBuffer);
+        }
+
+        private Controller GetControllerSlow(ITextView textView, ITextBuffer subjectBuffer)
+        {
+            AssertIsForeground();
+
+            return textView.GetOrCreatePerSubjectBufferProperty(
+                subjectBuffer,
+                s_controllerPropertyKey,
+                (textView, subjectBuffer) => new Controller(
+                    ThreadingContext,
+                    textView,
+                    subjectBuffer,
+                    _signatureHelpPresenter,
+                    _listener,
+                    new DocumentProvider(ThreadingContext),
+                    _signatureHelpProviders,
+                    _completionBroker));
+        }
```

Comment:
was this all basically moved down without any other significant change?

Reply:
The intent is yes, but I can verify.

Reply:
i trust you.  i just like to know with 'code moves' if it's effectively substantively unchanged or not.  my presumption is that it is unchanged.  but if i'm wrong, it's good to know to review more carefully.

Reply:
Yes, this is moved from `AbstractSignatureHelpCommandHandler` and `Controller`.